### PR TITLE
data_source_client: basic client data source

### PR DESCRIPTION
**Pull request by harryrose**
The beginnings of a data source to fetch client information from auth0.  Only a subset of fields are supported in this PR as it's not clear how many will be useful for terraform.  Certainly these fields are useful for my purposes at least...

Note that these changes are dependent upon github.com/yieldr/go-auth0/pull/21

### CODE REVIEW

**Comment by alexkappa** [comment]
Hi @harryrose , thanks for submitting this PR! I'll try to review this as soon as I can.

**Comment by harryrose** [comment]
HI @alexkappa. thanks for the comments. I'll try to make the changes this afternoon.

File: auth0/data_source_auth0_client.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/yieldr/go-auth0/management"
+)
+
+const (
+	clientDataSchemaName              = "name"
+	clientDataSchemaDescription       = "description"
+	clientDataSchemaClientID          = "client_id"
+	clientDataSchemaClientSecret      = "client_secret"
+	clientDataSchemaSigningKeys       = "signing_keys"
```
[the remark this answers is not in the document]
Comment (harryrose) [reply auth0/data_source_auth0_client.go]: so I'm clear, do you mean the use of these constants?

[the remark this answers is not in the document]
Comment (alexkappa) [reply auth0/data_source_auth0_client.go]: Yup

File: auth0/data_source_auth0_client.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+				Computed: true,
+			},
+		},
+	}
+}
+
+func readClientDataSource(r *schema.ResourceData, m interface{}) error {
+	api := m.(*management.Management)
+
+	clientID, ok := r.Get(clientDataSchemaClientID).(string)
```
[the remark this answers is not in the document]
Comment (harryrose) [reply auth0/data_source_auth0_client.go]: cool. switched to String, and modified the check to check for nil.  Like you say, terraform should enforce it, but it seems wrong to deference a pointer without checking it.

**Comment by harryrose** [comment]
First off, sorry for the delay in getting back to you on this.

I'm just looking at writing tests, but struggling to find a way of achieving it without writing a mock Auth0 HTTP API.  Ideally I'd like to provide a mock Management instance, but that doesn't seem possible without writing a local wrapper or modifying the auth0 client library.

I've looked at how other tests were written, but it seems that they're not working:
e.g. `make testacc` returns

```
--- FAIL: TestAccClientGrant (0.01s)
    testing.go:527: Step 0 error: config is invalid: 3 problems:
        
        - provider.auth0: "client_id": required field is not set
        - provider.auth0: "client_secret": required field is not set
        - provider.auth0: "domain": required field is not set
```
In my test so far, I've filled in dummy values, but of course the provider is attempting to make HTTP requests to those dummy values.  

My initial thoughts are to write a wrapper that provides methods to accessing the individual management clients interfaces; something along the lines of:

```
type managementWrapper struct {
 *management.Management
}

func (mw *managementWrapper) ClientManager() ClientManager {
  return mw.Client
}

type ClientManager interface {
 ... methods from management.Client that we're using locally
}
```

In the provider methods, we can then convert to the interface types containing the methods we're specifically interested in:

```
func readClientDataSource(r *schema.ResourceData, m interface{}) error {
	api := m.(interface{
          ClientManager() ClientManager
        })
 ...
```
which in turn would allow us to provide a mock management interface, allowing us to test without requiring a live auth0 api.

What are your thoughts on this?

**Pull request closed without merging**